// !WITH_NEW_INFERENCE

interface T {
  val a = <!NI;PROPERTY_INITIALIZER_IN_INTERFACE!><!PROPERTY_INITIALIZER_IN_INTERFACE!><!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>Foo<!><!>.<!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>bar<!><!>()<!><!>
}