// !DIAGNOSTICS: -UNREACHABLE_CODE
// !WITH_NEW_INFERENCE

fun foo() {
    val <!UNUSED_VARIABLE!>text<!>: List<Any> = null!!
    text.<!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>map<!><!> <!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>Any<!><!><!NI;SYNTAX!><!SYNTAX!>?<!><!>::<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>toString<!><!>
}
