// !WITH_NEW_INFERENCE

val unwrapped = <!NI;UNRESOLVED_REFERENCE!><!UNRESOLVED_REFERENCE!>some<!><!>.<!NI;SYNTAX!><!SYNTAX!><<!><!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>cabc<!><!><!NI;SYNTAX!><!SYNTAX!>$Wrapper<!><!><out <!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>Any<!><!>>::<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>unwrap<!><!>