// !WITH_NEW_INFERENCE

open class T {
    val x : Int? = null
}

class A {
    companion object: T() {
    }
}

class B {
    companion object: T() {
    }
}

fun test() {
    if (A.x != null) {
        useInt(<!NI;DEBUG_INFO_SMARTCAST!><!DEBUG_INFO_SMARTCAST!>A.x<!><!>)
        useInt(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>B.x<!><!>)
    }
}

fun useInt(i: Int) = i