// !WITH_NEW_INFERENCE

annotation class A1
annotation class A2(val some: Int = 12)

fun <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A1<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A2(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A1(<!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>12<!><!>)<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2(<!NI;TYPE_MISMATCH!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>"Test"<!><!><!>)<!><!><!><!>  T> topFun() = 12

class SomeClass {
    fun <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A1<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A2(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A1(<!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>12<!><!>)<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2(<!NI;TYPE_MISMATCH!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>"Test"<!><!><!>)<!><!><!><!> T> method() = 12

    fun foo() {
        fun <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A1<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A2(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A1(<!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>12<!><!>)<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2(<!NI;TYPE_MISMATCH!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>"Test"<!><!><!>)<!><!><!><!>  T> innerFun() = 12
    }
}