// !DIAGNOSTICS: -UNUSED_VARIABLE
// !WITH_NEW_INFERENCE
// FILE: A.java

import java.util.*;

public class A {
    void foo(List<? super String> x) {}
    void bar(List<? super Object> x) {}
}
// FILE: main.kt

fun main(a: A, ml: MutableList<String>, l: List<String>) {
    a.foo(ml)
    a.foo(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>l<!><!>)

    a.bar(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>ml<!><!>)
    a.bar(<!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>l<!><!>)
}
